### PR TITLE
Add the RLS as a submodule and build a package out of it

### PR DESCRIPTION
**Pull request by nrc**
r? @brson (and cc @alexcrichton) Please review closely, I am not at all convinced I've done the right things here. I did run `x.py dist` and it makes an rls package which looks right to my eyes, but I haven't tested on non-linux platforms nor am I really sure what it should look like.

This does not attempt to run tests for the RLS yet.

### CODE REVIEW

**Comment by alexcrichton** [comment]
A few thoughts:

* this is touching the windows packaging code. That configuration hasn't been touched in what I believe is literally years. I'd highly recommend running locally to confirm the installers actually work (I'd rather not discover this after the nightly is published)
* I thought we were gonna fold rust-analysis into the rls? (in terms of packages) That way we don't have to deal with the problem of encoding package dependencies.
* I'd also recommend running the OSX installers to ensure they work
* I think that this is going to install the rls by default with the combined installers, is that what we want? (I thought we wanted it optional)
* This is likely also going to want to update `build-manifest/src/main.rs` I believe
* @brson has indicated that rustup needs support for the RLS, so we may want to wait on that?

File: src/bootstrap/lib.rs [hunk 1]
```diff
@@ -1018,6 +1018,21 @@ impl Build {
         panic!("failed to find version in cargo's Cargo.toml")
     }
 
+    /// Returns the `a.b.c` version that the RLS is at.
+    fn rls_release_num(&self) -> String {
```
Comment (alexcrichton) [inline src/bootstrap/lib.rs]: Could this be deduplicated with the cargo version of this function above?

**Comment by alexcrichton** [comment]
I think it'd also be helpful to generate some rls tarballs and poke around to make sure they look right (e.g. similar to Cargo tarballs and such)

**Comment by brson** [comment]
> I thought we were gonna fold rust-analysis into the rls? (in terms of packages) That way we don't have to deal with the problem of encoding package dependencies.

I think we should not do that after all, the reason being that the rls also depends on the rust-src package, and there are reasons not to merge rust-src into rls, so if we have to have 2 packages, why not 3.

**Comment by alexcrichton** [comment]
ah ok yeah that makes sense, if we have rust-src + rls at least then it doesn't make much sense to worry about rust-analysis.

File: src/bootstrap/dist.rs [hunk 2]
```diff
@@ -598,6 +636,11 @@ pub fn extended(build: &Build, stage: u32, target: &str) {
     let cargo_installer = dist.join(format!("cargo-{}-{}.tar.gz",
                                             build.package_vers(&cargo_vers),
                                             target));
+    let rls_installer = dist.join(format!("{}.tar.gz",
+                                          pkgname(build, "rls")));
```
Comment (brson) [inline src/bootstrap/dist.rs]: This needs to be "{}-{}.tar.gz", with a final argument of `target`. That will put the target triple in the tarball filename, which is needed for disambiguation.

File: src/bootstrap/dist.rs [hunk 3]
```diff
@@ -669,20 +714,28 @@ pub fn extended(build: &Build, stage: u32, target: &str) {
...
+        t!(fs::create_dir_all(pkg.join("rust-analysis")));
         t!(fs::create_dir_all(pkg.join("rust-docs")));
         t!(fs::create_dir_all(pkg.join("rust-std")));
 
         cp_r(&work.join(&format!("{}-{}", pkgname(build, "rustc"), target)),
              &pkg.join("rustc"));
         cp_r(&work.join(&format!("cargo-nightly-{}", target)),
              &pkg.join("cargo"));
+        cp_r(&work.join(pkgname(build, "rls")),
+             &pkg.join("rls"));
```
Comment (brson) [inline src/bootstrap/dist.rs]: This also needs the target.

File: src/bootstrap/dist.rs [hunk 4]
```diff
@@ -729,6 +786,12 @@ pub fn extended(build: &Build, stage: u32, target: &str) {
         cp_r(&work.join(&format!("cargo-nightly-{}", target))
                   .join("cargo"),
              &exe.join("cargo"));
+        cp_r(&work.join(pkgname(build, "rls"))
+                  .join("rls"),
+             &exe.join("rls"));
```
Comment (brson) [inline src/bootstrap/dist.rs]: And here.

File: src/bootstrap/dist.rs [hunk 5]
```diff
@@ -669,20 +714,28 @@ pub fn extended(build: &Build, stage: u32, target: &str) {
         let _ = fs::remove_dir_all(&pkg);
         t!(fs::create_dir_all(pkg.join("rustc")));
         t!(fs::create_dir_all(pkg.join("cargo")));
+        t!(fs::create_dir_all(pkg.join("rls")));
+        t!(fs::create_dir_all(pkg.join("rust-analysis")));
```
Comment (brson) [inline src/bootstrap/dist.rs]: All of the changes in this file from here down only affect the platform-specific .msi and .pkg installers, and will not work correctly without further modifications of their definitions. I think you should remove them all for now (everything inside the `target.contains("apple-darwin")` and `target.contains("windows").

File: src/bootstrap/dist.rs (hunk 3 above)
Comment (brson) [inline src/bootstrap/dist.rs]: (but just delete it)

File: src/bootstrap/dist.rs [hunk 6]
```diff
@@ -797,6 +862,26 @@ pub fn extended(build: &Build, stage: u32, target: &str) {
...
+                        .arg("rust-analysis")
+                        .args(&heat_flags)
+                        .arg("-cg").arg("AnalysisGroup")
+                        .arg("-dr").arg("Analysis")
+                        .arg("-var").arg("var.AnalysisDir")
+                        .arg("-out").arg(exe.join("AnalysisGroup.wxs"))
+                        .arg("-t").arg(etc.join("msi/squash-components.xsl")));
+        build.run(Command::new(&heat)
+                        .current_dir(&exe)
+                        .arg("dir")
```
Comment (brson) [inline src/bootstrap/dist.rs]: This is all .msi stuff and isn't going to work without defining RlsGroup.wxs etc. Just remove for now.

**Comment by brson** [comment]
As @alexcrichton said build-manifest needs to be taught about the rls in order for rustup to know about it.

**Comment by nrc** [comment]
@alexcrichton 

> I think that this is going to install the rls by default with the combined installers, is that what we want? (I thought we wanted it optional)

Could you explain why please? My understanding was that by adding only to `extended` it becomes an optional component.

**Comment by alexcrichton** [comment]
@nrc oh that's for rustup, but the combined installers (e.g. `foo.pkg`, `foo.msi`, etc) I think have a different configuration for what's optional and what isn't.

**Comment by nrc** [comment]
> oh that's for rustup, but the combined installers (e.g. foo.pkg, foo.msi, etc) I think have a different configuration for what's optional and what isn't.

So, I think I've removed all that stuff (at brson's request) and that it should be just optional now?

Ready for re-review @alexcrichton /@brson

**Comment by alexcrichton** [comment]
@bors: r-


Oh I was hoping to do some more testing on this before landing, have we run the installers on Windows yet? I also am worried about the regression to cycle time this will bring because we are already a half hour too slow.

To test I think we should generate the windows and osx installers and make sure they look right. For cycle time we should split all the dist builders in two (they're all producing two targets right now). 

Sorry on my phone so it's hard to link issues but I'll try to fill in soon. I'll also try to help out with this when I'm back from traveling, but that'll be a week or so

**Comment by alexcrichton** [comment]
I've sent [a PR](https://github.com/rust-lang/rust/pull/40998) to split the dist builders to help the cycle time issue. I won't have access to a Windows machine for another week, but if you test ahead of time that'd also be fine.

**Comment by nrc** [comment]
I've tested this on Linux and @jonathandturner has tested on OS X both work and both generate the tar files we'd expect to see. I don't think we have a Windows environment we can test on though.

**Comment by alexcrichton** [comment]
Thanks @nrc!

Can you also try running the final step, `./x.py dist hash-and-sign` and see if that works? That's the final "build manifest" stage which generates the toml files that we distribute. It's configured from rust-central-station [here](https://github.com/alexcrichton/rust-central-station/blob/784b1137c9abe858181e186528aef87bffc49f80/promote-release/src/main.rs#L158-L183) and called [here](https://github.com/alexcrichton/rust-central-station/blob/784b1137c9abe858181e186528aef87bffc49f80/promote-release/src/main.rs#L262-L268). 

I think it should work with lots of artifacts missing, you'll just have some blanks.

**Comment by jonathandturner** [comment]
@alexcrichton does "failed to specify `dist.sign-folder` in `config.toml`" count as lots of artefacts missing?

**Comment by alexcrichton** [comment]
Oh that's the [configuration step](https://github.com/alexcrichton/rust-central-station/blob/784b1137c9abe858181e186528aef87bffc49f80/promote-release/src/main.rs#L158-L183), you'll need to configure a few things in `config.toml` for it to work.

**Review by alexcrichton — COMMENTED** [review]
Ok got home and have dome some investigation. I found one showstopper on Windows so far (commented below, not specific to Windows), but some other observations are:

* The combined tarball installer includes rls/rust-analsyis
* Other combined installers (pkg/msi/exe) do not contain rls/rust-analysis


I'm running the `hash-and-sign` step right now to see what happens.

**Comment by alexcrichton** [comment]
It looks like `build-manifest` also lists `available = false` for the rls for all platforms.

[This likely needs an update](https://github.com/rust-lang/rust/blob/8493dd6d6e8f5e1280ae3dc05135a7e2897c36ac/src/tools/build-manifest/src/main.rs#L341-L349).

**Comment by alexcrichton** [comment]
(note that that's also not windows specific)

**Comment by alexcrichton** [comment]
A few other things discovered along the way:

* The RLS uses a different version of Cargo than is vendored in the repository
* `rls --version` does not work, it presumably starts up the server
* We don't run RLS tests

**Comment by alexcrichton** [comment]
@bors: r+

I'll open follow-up tracking issues for various issues

**Comment by aidanhs** [comment]
```
DEBUG:sccache::simples3::s3: GET http://rust-lang-ci-sccache.s3.amazonaws.com/2/6/1/26128c6d17efd88aaf5f6eec4f07d42d4a697959
WARN:sccache::cache::s3: Got AWS error: Error(BadHTTPStatus(NotFound), State { next_error: None })
DEBUG:sccache::compiler::compiler: [CheckSymbolExists.c.o]: Cache miss
DEBUG:sccache::server: [Error(ProcessError(Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c: In function \'main\':\n/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:8:19: error: \'arc4random\' undeclared (first use in this function)\n   return ((int*)(&arc4random))[argc];\n                   ^\n/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:8:19: note: each undeclared identifier is reported only once for each function it appears in\n" }), State { next_error: None })] compilation failed: "CheckSymbolExists.c.o"
ERROR:sccache::server: [Error(ProcessError(Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c: In function \'main\':\n/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:8:19: error: \'arc4random\' undeclared (first use in this function)\n   return ((int*)(&arc4random))[argc];\n                   ^\n/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:8:19: note: each undeclared identifier is reported only once for each function it appears in\n" }), State { next_error: None })] 	CheckSymbolExists.c.o
```
Sounds valid?

**Comment by jonathandturner** [comment]
@aidanhs - not sure how this would be related to the RLS stuff, though perhaps it's related to other infra fixes?

**Comment by alexcrichton** [comment]
Yes this is a valid error, but not related to sccache:

```
[00:45:06]     Finished release [optimized] target(s) in 15.34 secs

[00:45:06] Copying stage2 test from stage2 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / asmjs-unknown-emscripten)

[00:45:06] Dist std stage2 (x86_64-unknown-linux-gnu -> asmjs-unknown-emscripten)

[00:45:06] gen-installer: looking for programs

[00:45:06] gen-installer: 

[00:45:06] gen-installer: found tar

[00:45:06] gen-installer: found cp

[00:45:06] gen-installer: found rm

[00:45:06] gen-installer: found mkdir

[00:45:06] gen-installer: found echo

[00:45:06] gen-installer: found tr

[00:45:06] gen-installer: found awk

[00:45:06] gen-installer: 

[00:45:06] gen-installer: processing arguments

[00:45:06] gen-installer: 

[00:45:06] gen-installer: CFG_PRODUCT_NAME     := Rust 

[00:45:06] gen-installer: CFG_COMPONENT_NAME   := rust-std-asmjs-unknown-emscripten 

[00:45:06] gen-installer: CFG_PACKAGE_NAME     := rust-std-nightly-asmjs-unknown-emsc ...

[00:45:06] gen-installer: CFG_REL_MANIFEST_DIR := rustlib 

[00:45:06] gen-installer: CFG_SUCCESS_MESSAGE  := std-is-standing-at-the-ready. 

[00:45:06] gen-installer: CFG_LEGACY_MANIFEST_DIRS := rustlib,cargo 

[00:45:06] gen-installer: CFG_NON_INSTALLED_OVERLAY :=  

[00:45:06] gen-installer: CFG_BULK_DIRS        :=  

[00:45:06] gen-installer: CFG_IMAGE_DIR        := /checkout/obj/build/tmp/dist/rust-s ...

[00:45:06] gen-installer: CFG_WORK_DIR         := /checkout/obj/build/tmp/dist 

[00:45:06] gen-installer: CFG_OUTPUT_DIR       := /checkout/obj/build/dist 

[00:45:06] gen-installer: 

[00:45:06] gen-installer: validating arguments

[00:45:06] gen-installer: 

[00:45:06] gen-install-script: looking for install programs

[00:45:06] gen-install-script: 

[00:45:06] gen-install-script: found sed

[00:45:06] gen-install-script: found chmod

[00:45:06] gen-install-script: found cat

[00:45:06] gen-install-script: 

[00:45:06] gen-install-script: processing arguments

[00:45:06] gen-install-script: 

[00:45:06] gen-install-script: CFG_PRODUCT_NAME     := Rust 

[00:45:06] gen-install-script: CFG_REL_MANIFEST_DIR := rustlib 

[00:45:06] gen-install-script: CFG_SUCCESS_MESSAGE  := std-is-standing-at-the-ready. 

[00:45:06] gen-install-script: CFG_OUTPUT_SCRIPT    := /checkout/obj/build/tmp/dist/rust-s ...

[00:45:06] gen-install-script: CFG_LEGACY_MANIFEST_DIRS := rustlib,cargo 

[00:45:06] gen-install-script: 

[00:45:06] gen-install-script: validating arguments

[00:45:06] gen-install-script: 

[00:45:08] thread 'main' panicked at 'assertion failed: build.config.extended', /checkout/src/bootstrap/dist.rs:320

[00:45:08] note: Run with `RUST_BACKTRACE=1` for a backtrace.

[00:45:08] Build completed unsuccessfully in 0:44:05
```

I'll try to fix soon.

**Comment by frewsxcv** [comment]
@bors retry 

Prioritizing the rollup which includes this

**Pull request merged**